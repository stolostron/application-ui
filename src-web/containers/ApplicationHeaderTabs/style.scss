/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure 
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

#applicationheadertabs {
  .whiteSpacer {
    background-color: white;
    height: 40px;
    padding-left: 5%;
    padding-right: 5%;
  }

  .some-class {
    z-index: 1;
  }

  .page-content-container {
    padding-right: 0px;
    padding-left: 0px;
  }

  .tech-preview-tag {
    background-color: #d93f00;
    width: 125px;
    height: 26px;
    line-height: 22px;
    border-radius: 3px;
    text-align: center;
    position: absolute;
    right: 5%;
    z-index: 10;

    .bx--tooltip__trigger {
      font-family: RedHatText;
      font-size: 12px;
      color: white;
      border: none;
      cursor: pointer;
    }

    #definition-tooltip-1 {
      background-color: #393939;
      width: 256px;
      max-width: none;
      border-radius: 3px;
      margin-top: 18px;
      padding: 12px;
      right: 0;
      z-index: 10;
    }

    .bx--tooltip__caret {
      background-color: #393939;
      margin-left: 0;
      left: 228px;
    }

    #definition-tooltip-1 > p {
      font-family: RedHatText;
      font-size: 14px;
      text-align: left;
    }
  }
}
