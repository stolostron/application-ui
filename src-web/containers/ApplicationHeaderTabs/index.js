/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2017, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

// seems to be an issue with this rule and redux
/* eslint-disable import/no-named-as-default */

import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as Actions from '../../actions'
import loadable from 'loadable-components'
import { updateSecondaryHeader ,
  delResourceSuccessFinished,
  mutateResourceSuccessFinished
} from '../../actions/common'
import { Tabs, Tab } from 'carbon-components-react'
import msgs from '../../../nls/platform.properties'
import { withLocale } from '../../providers/LocaleProvider'
import resources from '../../../lib/shared/resources'
import _ from 'lodash'

import { RESOURCE_TYPES } from '../../../lib/shared/constants'

resources(() => {
  require('./style.scss')
})

export const ApplicationsTab = loadable(() =>
  import(/* webpackChunkName: "applications" */ '../ApplicationsTab')
)

export const ApplicationDeploymentPipeline = loadable(() =>
  import(/* webpackChunkName: "applicationdeploymentpipeline" */ '../../components/ApplicationDeploymentPipeline')
)

export const ApplicationCreationPage = loadable(() =>
  import(/* webpackChunkName: "applicationcreatepage" */ '../../components/ApplicationCreationPage/ApplicationCreationPage')
)

const routes = ['', 'advanced', 'yaml', 'create']
// This will render the two tabs
// Overview, Resources
const ApplicationHeaderTabs = withLocale(
  ({
    params,
    locale,
    serverProps,
    mutateSuccessFinished,
    deleteSuccessFinished,
    updateSecondary
  }) => {

    // process restful api into which tab to show
    const {history, location} = params
    const pathname = _.get(location, 'pathname', '')
    const segments = pathname.split('/')
    const isSingleApplicationView = segments.length>=5
    let route = ''
    if (segments.length===4 || segments.length===6) {
      route = segments.pop()
    }
    const basePath = segments.join('/')
    const selectedTab = routes.indexOf(route)
    const selectedAppName = isSingleApplicationView ? segments.pop() : null
    const selectedAppNamespace = isSingleApplicationView ? segments.pop() : null
    const selectedApp = {
      isSingleApplicationView,
      selectedAppName,
      selectedAppNamespace
    }

    // update breadcrumbs
    const breadcrumbs = []
    breadcrumbs.push({
      label: 'Applications',
      url: segments.join('/')
    })
    if (isSingleApplicationView) {
      breadcrumbs.push({
        label: selectedAppName,
        url: [...segments, selectedAppNamespace, selectedAppName].join('/')
      })
    }
    updateSecondary(selectedAppName||'Applications', breadcrumbs)
    selectedApp.breadcrumbs = breadcrumbs

    const noop = () => {
      // no op function for optional properties
    }

    const renderTab = thisTab => {
      if (selectedTab === thisTab) {
        switch (thisTab) {
        default:
        case 0:
          return (
            <div className="some-content">
              <ApplicationsTab
                secondaryHeaderProps={{ title: 'routes.applications' }}
                />
            </div>
          )
        case 1:
          return (
            <div className="page-content-container">
              <ApplicationDeploymentPipeline serverProps={serverProps} selectedApp={selectedApp} />
            </div>
          )
        case 2:
          return (
            <div className="page-content-container">
              <ApplicationCreationPage serverProps={serverProps} editApplication={selectedApp} />
            </div>
          )
        }
      }
      return null
    }

    if (selectedTab<=2) {
      return (
        <div id="applicationheadertabs">
          <div className="whiteSpacer">
            <Tabs
              className="some-class"
              onClick={noop}
              onKeyDown={noop}
              selected={selectedTab}
              onSelectionChange={id => {
                // Remove previous success message if any
                mutateSuccessFinished(RESOURCE_TYPES.HCM_CHANNELS)
                mutateSuccessFinished(RESOURCE_TYPES.HCM_SUBSCRIPTIONS)
                mutateSuccessFinished(RESOURCE_TYPES.HCM_PLACEMENT_RULES)
                mutateSuccessFinished(RESOURCE_TYPES.QUERY_APPLICATIONS)
                deleteSuccessFinished(RESOURCE_TYPES.HCM_CHANNELS)
                deleteSuccessFinished(RESOURCE_TYPES.HCM_SUBSCRIPTIONS)
                deleteSuccessFinished(RESOURCE_TYPES.HCM_PLACEMENT_RULES)
                deleteSuccessFinished(RESOURCE_TYPES.QUERY_APPLICATIONS)
                // open the tab
                const route = id>0 ? `${routes[id]}`:''
                history.push(`${basePath}/${route}`)
              }}
              tabcontentclassname="tab-content"
              >
              <Tab
                disabled={false}
                onClick={noop}
                onKeyDown={noop}
                label={msgs.get('description.title.overview', locale)}
                >
                {renderTab(0)}
              </Tab>
              <Tab
                disabled={false}
                onClick={noop}
                onKeyDown={noop}
                label={msgs.get('description.title.deployments', locale)}
                >
                {renderTab(1)}
              </Tab>
              {isSingleApplicationView&&<Tab
                disabled={false}
                onClick={noop}
                onKeyDown={noop}
                label={msgs.get('description.title.yaml', locale)}
              >
                  {renderTab(2)}
                </Tab>}
            </Tabs>
          </div>
        </div>
      )
    } else {
      return <ApplicationCreationPage secondaryHeaderProps={{title: 'application.create.title'}} />
    }
  }
)

const mapDispatchToProps = dispatch => {
  return {
    actions: bindActionCreators(Actions, dispatch),
    mutateSuccessFinished: resourceType =>
      dispatch(mutateResourceSuccessFinished(resourceType)),
    deleteSuccessFinished: resourceType =>
      dispatch(delResourceSuccessFinished(resourceType)),
    updateSecondary: (title, breadcrumbs) =>
      dispatch(updateSecondaryHeader(title, [], breadcrumbs))
  }
}
const mapStateToProps = state => {
  const { role, AppOverview } = state
  return {
    userRole: role && role.role,
    selectedAppTab: AppOverview.selectedAppTab
  }
}

export default withLocale(
  connect(mapStateToProps, mapDispatchToProps)(ApplicationHeaderTabs)
)
