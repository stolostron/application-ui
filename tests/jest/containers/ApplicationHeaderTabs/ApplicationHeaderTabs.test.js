/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

const React = require("../../../../node_modules/react");

import ApplicationHeaderTabs from "../../../../src-web/containers/ApplicationHeaderTabs";

import { mount } from "enzyme";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";

import configureMockStore from "redux-mock-store";

import {
  reduxStoreAppPipelineWithCEM,
  serverProps
} from "../../components/TestingData";

const mockStore = configureMockStore();

describe("ApplicationHeaderTabs", () => {
  it("has functioning onclick, one app", () => {
    const storeApp = mockStore(reduxStoreAppPipelineWithCEM);

    const component = mount(
      <Provider store={storeApp}>
        <ApplicationHeaderTabs
          showSingleApp={false}
          serverProps={serverProps}
        />
      </Provider>
    );
  });

  it("ApplicationHeaderTabs renders correctly.", () => {
    reduxStoreAppPipelineWithCEM.AppOverview.selectedAppTab = 2;
    const storeApp = mockStore(reduxStoreAppPipelineWithCEM);

    const tree = renderer
      .create(
        <Provider store={storeApp}>
          <ApplicationHeaderTabs
            showSingleApp={false}
            serverProps={serverProps}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ApplicationHeaderTabs renders correctly.", () => {
    reduxStoreAppPipelineWithCEM.AppOverview.selectedAppTab = 1;
    const storeApp = mockStore(reduxStoreAppPipelineWithCEM);

    const tree = renderer
      .create(
        <Provider store={storeApp}>
          <ApplicationHeaderTabs
            showSingleApp={true}
            serverProps={serverProps}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ApplicationHeaderTabs renders correctly.", () => {
    reduxStoreAppPipelineWithCEM.AppOverview.selectedAppTab = 2;
    const storeApp = mockStore(reduxStoreAppPipelineWithCEM);

    const tree = renderer
      .create(
        <Provider store={storeApp}>
          <ApplicationHeaderTabs
            showSingleApp={true}
            serverProps={serverProps}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
