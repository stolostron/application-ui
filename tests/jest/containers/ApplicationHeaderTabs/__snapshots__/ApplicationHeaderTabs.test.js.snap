// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationHeaderTabs ApplicationHeaderTabs renders correctly. 1`] = `
<div
  id="applicationheadertabs"
>
  <div
    className="whiteSpacer"
  >
    <nav
      className="bx--tabs some-class"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="navigation"
      selected={0}
      tabcontentclassname="tab-content"
    >
      <div
        aria-label="listbox"
        className="bx--tabs-trigger"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <a
          className="bx--tabs-trigger-text"
          href="#"
          onClick={[Function]}
          tabIndex={-1}
        >
          Overview
        </a>
        <svg
          alt="show menu options"
          aria-label="show menu options"
          fillRule="evenodd"
          height="5"
          role="img"
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            show menu options
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
      <ul
        className="bx--tabs__nav bx--tabs__nav--hidden"
        role="tablist"
      >
        <li
          className="bx--tabs__nav-item bx--tabs__nav-item--selected"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={true}
          tabIndex={-1}
        >
          <a
            aria-selected={true}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Overview
          </a>
        </li>
        <li
          className="bx--tabs__nav-item"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={false}
          tabIndex={-1}
        >
          <a
            aria-selected={false}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Advanced Configuration
          </a>
        </li>
      </ul>
    </nav>
    <div
      aria-hidden={false}
      className="tab-content"
      hidden={false}
      selected={true}
    >
      <div
        className="some-content"
      />
    </div>
    <div
      aria-hidden={true}
      className="tab-content"
      hidden={true}
      selected={false}
    />
  </div>
</div>
`;

exports[`ApplicationHeaderTabs ApplicationHeaderTabs renders correctly. 2`] = `
<div
  id="applicationheadertabs"
>
  <div
    className="whiteSpacer"
  >
    <nav
      className="bx--tabs some-class"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="navigation"
      selected={0}
      tabcontentclassname="tab-content"
    >
      <div
        aria-label="listbox"
        className="bx--tabs-trigger"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <a
          className="bx--tabs-trigger-text"
          href="#"
          onClick={[Function]}
          tabIndex={-1}
        >
          Overview
        </a>
        <svg
          alt="show menu options"
          aria-label="show menu options"
          fillRule="evenodd"
          height="5"
          role="img"
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            show menu options
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
      <ul
        className="bx--tabs__nav bx--tabs__nav--hidden"
        role="tablist"
      >
        <li
          className="bx--tabs__nav-item bx--tabs__nav-item--selected"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={true}
          tabIndex={-1}
        >
          <a
            aria-selected={true}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Overview
          </a>
        </li>
        <li
          className="bx--tabs__nav-item"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={false}
          tabIndex={-1}
        >
          <a
            aria-selected={false}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Advanced Configuration
          </a>
        </li>
      </ul>
    </nav>
    <div
      aria-hidden={false}
      className="tab-content"
      hidden={false}
      selected={true}
    >
      <div
        className="some-content"
      />
    </div>
    <div
      aria-hidden={true}
      className="tab-content"
      hidden={true}
      selected={false}
    />
  </div>
</div>
`;

exports[`ApplicationHeaderTabs ApplicationHeaderTabs renders correctly. 3`] = `
<div
  id="applicationheadertabs"
>
  <div
    className="whiteSpacer"
  >
    <nav
      className="bx--tabs some-class"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="navigation"
      selected={0}
      tabcontentclassname="tab-content"
    >
      <div
        aria-label="listbox"
        className="bx--tabs-trigger"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <a
          className="bx--tabs-trigger-text"
          href="#"
          onClick={[Function]}
          tabIndex={-1}
        >
          Overview
        </a>
        <svg
          alt="show menu options"
          aria-label="show menu options"
          fillRule="evenodd"
          height="5"
          role="img"
          viewBox="0 0 10 5"
          width="10"
        >
          <title>
            show menu options
          </title>
          <path
            d="M0 0l5 4.998L10 0z"
          />
        </svg>
      </div>
      <ul
        className="bx--tabs__nav bx--tabs__nav--hidden"
        role="tablist"
      >
        <li
          className="bx--tabs__nav-item bx--tabs__nav-item--selected"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={true}
          tabIndex={-1}
        >
          <a
            aria-selected={true}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Overview
          </a>
        </li>
        <li
          className="bx--tabs__nav-item"
          disabled={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="presentation"
          selected={false}
          tabIndex={-1}
        >
          <a
            aria-selected={false}
            className="bx--tabs__nav-link"
            href="#"
            role="tab"
            tabIndex={0}
          >
            Advanced Configuration
          </a>
        </li>
      </ul>
    </nav>
    <div
      aria-hidden={false}
      className="tab-content"
      hidden={false}
      selected={true}
    >
      <div
        className="some-content"
      />
    </div>
    <div
      aria-hidden={true}
      className="tab-content"
      hidden={true}
      selected={false}
    />
  </div>
</div>
`;
